src/uarch-replay-reset-state-access.h

```diff
@@ -0,0 +1,234 @@
+                throw std::invalid_argument{"hash of read data and read hash at access " +
+                    std::to_string(access_to_report()) + " does not match read hash"};
+            }
+        }
+        if (!access.get_written_hash().has_value()) {
+            throw std::invalid_argument{"write access " + std::to_string(access_to_report()) + " has no written hash"};
+        }
+        if (access.get_written().has_value()) {
+            // if written data is available then its hash and the logged written hash must match
+            hash_type computed_hash;
+            get_hash(hasher, access.get_written().value().data(), access.get_written().value().size(), computed_hash);
+            if (computed_hash != access.get_written_hash().value()) {
+                throw std::invalid_argument{"hash of written data and written hash at access " +
+                    std::to_string(access_to_report()) + " does not match written hash"};
+            }
+        }
+        if (m_verify_proofs) {
+            if (!access.get_proof().has_value()) {
+                throw std::invalid_argument{"write access " + std::to_string(access_to_report()) + " has no proof"};
+            }
+            const auto &proof = access.get_proof().value(); // NOLINT(bugprone-unchecked-optional-access)
+            if (proof.get_target_address() != access.get_address()) {
+                throw std::invalid_argument{"mismatch in write access " + std::to_string(access_to_report()) +
+                    " address and its proof address"};
+            }
+            if (m_root_hash != proof.get_root_hash()) {
+                throw std::invalid_argument{
+                    "mismatch in write access " + std::to_string(access_to_report()) + " root hash"};
+            }
+            auto rolling_hash = access.get_read_hash();
+            if (rolling_hash != proof.get_target_hash()) {
+                throw std::invalid_argument{
+                    "value before write access " + std::to_string(access_to_report()) + " does not match target hash"};
+            }
+            roll_hash_up_tree(m_hasher, proof, rolling_hash);
+            if (rolling_hash != proof.get_root_hash()) {
+                throw std::invalid_argument{
+                    "value before write access " + std::to_string(access_to_report()) + " fails proof"};
+            }
+            // computes the new root hash by rolling up the write hash
+            m_root_hash = access.get_written_hash().value();
+            roll_hash_up_tree(m_hasher, proof, m_root_hash);
+        }
+        if (access.get_written_hash().value() != uarch_pristine_state_hash) {
+            throw std::invalid_argument{"expected written hash of access " + std::to_string(access_to_report()) +
+                " to be the start hash of the pristine uarch state"};
+        }
+        m_next_access++;
+    }
+
+    static void get_hash(hasher_type &hasher, const unsigned char *data, size_t len, hash_type &hash) {
```

Comment:
We use this construct so much that I'm thinking we should just add it to the `i-hasher` interface. What do you think? Then use that everywhere.